### PR TITLE
WIP: `is_valid_hom` and `enum_homs` functions

### LINKED ISSUE #36717: Possibility of extending `has_homomorphism_to` to support returning all homomorphisms?
As continuation of PR [#36504](https://github.com/sagemath/sage/pull/36504), I am working on some functionalities related to counting (graph) homomorphisms. Currently, Sage already has `has_homomorphism_to` function, but it only returns one homomorphism even if multiple exist.

It seems that this function is implemented with some LP solver. Would it be possible to extend it to support returning all homomorphisms?

**Edit**: This one will be addressed in a new branch.

### PR DESCRIPTION
**Pull request by guojing0**
This PR intends to resolve https://github.com/sagemath/sage/issues/36717 by introducing two functions:

1. `is_valid_hom(G, H, hom)` (Complete): Checking if `hom` is a valid homomorphism from `G` to `H`.
2. `enum_homs(G, H)` (**Work in progress**): Enumerating all homomorphisms from `G` to `H`.

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/sage/graphs/graph.py [hunk 1]
```diff
@@ -4476,6 +4476,82 @@ def is_factor_critical(self, matching=None, algorithm='Edmonds', solver=None, ve
         # The graph is factor critical if all vertices are marked even
         return len(even) == self.order()
 
+    def is_valid_hom(G, H, hom):
```
Comment (dcoudert) [inline src/sage/graphs/graph.py]: change the name to `is_homomorphism_to` or something like that. Avoid shortcuts like `hom`.

File: src/sage/graphs/graph.py [hunk 2]
```diff
@@ -4476,6 +4476,82 @@ def is_factor_critical(self, matching=None, algorithm='Edmonds', solver=None, ve
...
+    r"""
```
Comment (dcoudert) [inline src/sage/graphs/graph.py]: Please follow the coding style, in particular for indentation

File: src/sage/graphs/graph.py [hunk 3]
```diff
@@ -4476,6 +4476,82 @@ def is_factor_critical(self, matching=None, algorithm='Edmonds', solver=None, ve
...
+
+    INPUT:
+
+    - ``G`` -- a Sage graph
+
+    - ``H`` -- a Sage graph, as target of the homomorphism
+
+    - ``hom`` -- a graph homomorphism
+
+    OUTPUT:
```
Comment (dcoudert) [inline src/sage/graphs/graph.py]: no need to explain that the output is a boolean. It's obvious from the name of the method.

File: src/sage/graphs/graph.py [hunk 4]
```diff
@@ -4476,6 +4476,82 @@ def is_factor_critical(self, matching=None, algorithm='Edmonds', solver=None, ve
...
+        for bag_vtx in bag:
+            hom_pairs = []
+            for target_vtx in sorted_H:
+                hom_pairs.append({bag_vtx, target_vtx})
+
+            homs.append(hom_pairs)
+
+        return homs
+
+    # def enumerate_functions(vertices, g):
```
Comment (dcoudert) [inline src/sage/graphs/graph.py]: please don't had this kind of commented code in commits.

**Review by dcoudert — CHANGES_REQUESTED** [review]

File: src/sage/graphs/graph.py (hunk 2 above)
Comment (guojing0) [reply src/sage/graphs/graph.py]: For some reason copy-and-paste didn't work well.

File: src/sage/graphs/graph.py (hunk 4 above)
Comment (guojing0) [reply src/sage/graphs/graph.py]: I will keep in mind. I was mainly using it as reference and backup

File: src/sage/graphs/graph.py (hunk 3 above)
Comment (guojing0) [reply src/sage/graphs/graph.py]: I try to keep `INPUT:` and `OUTPUT:` consistent for each function, even if it's evident from the function name.

**Comment by dcoudert** [comment]
I'm not sure where you are going with this PR. It seems that you need many methods that should not be in `src/sage/graphs/graph.py` but in a new module `src/sage/graphs/homomorphisms.py` gathering methods related to homomorphisms.

In each method, make sure that you give enough details to explain what the method is about. The top single line description is often not enough to understand. You can add more text as we try to do in many methods of the graph module.

if you implement a method to enumerate homomorphisms, consider making an iterator rather than returning the list of all solutions. The number of homomorphisms might be large.

**Comment by guojing0** [comment]
> I'm not sure where you are going with this PR. It seems that you need many methods that should not be in `src/sage/graphs/graph.py` but in a new module `src/sage/graphs/homomorphisms.py` gathering methods related to homomorphisms.

[This](https://github.com/guojing0/sage/blob/hom-count/src/sage/graphs/hom_count.py) is what I have been working on, and checking valid hom and enumerating homs seem to be methods that could potentially be helpful to others.

If that's the case, I will consider making a new `homomorphism` folder or just `homomorphism.py(x)`.

> In each method, make sure that you give enough details to explain what the method is about. The top single line description is often not enough to understand. You can add more text as we try to do in many methods of the graph module.

I do intend to.

> if you implement a method to enumerate homomorphisms, consider making an iterator rather than returning the list of all solutions. The number of homomorphisms might be large.

Will do.

**Comment by dcoudert** [comment]
> [This](https://github.com/guojing0/sage/blob/hom-count/src/sage/graphs/hom_count.py) is what I have been working on, and checking valid hom and enumerating homs seem to be methods that could potentially be helpful to others.

OK.
The most important methods could be made available from `Graph`, but most of them could be in a dedicated module and interested users can always import these methods.

> If that's the case, I will consider making a new `homomorphism` folder or just `homomorphism.py(x)`.

A file is enough.

**Comment by guojing0** [comment]
>  The most important methods could be made available from Graph, but most of them could be in a dedicated module and interested users can always import these methods.

So you think we can close this PR as these two would be accessed via `hom.py`?

**Comment by dcoudert** [comment]
You can use this PR to put the code you plan to add when you are ready.

**Comment by guojing0** [comment]
> You can use this PR to put the code you plan to add when you are ready.

It's in another branch. So I probably would open a new PR based on that...

**Pull request closed without merging**